### DISCUSSION TITLE
Strongly typed translation params

### DISCUSSION THREAD

**Discussion by osdiab** [Xiphe/remix-polyglot#8] [Ideas]
Made this quick playground as food for thought for how template literal types could make for strongly typed i18n params (instead of a totally opaque string)

https://www.typescriptlang.org/play?#code/MYewdgzgLgBApmGBeGBvAUDGBzATnOKASzGwC4YAiACTgBs6QYBSVMAQwFs4BfAQhgB1ABbtYRCDACeIAK64YAM3YA3ELiJQ4LNnLA8A-JQA0mGKE4AHOkW5goFSgE05MUSu3sd7ACYArOGBiDx4YSzhcCHB2GygpPhMzOAAPYDouMSJwRwAtEBIAawgE9FD2SVBIKABudDq48JgABXZcLggAHgBJAEYADjAAZSgNUnhkrTAfSWhR7AA+ZBhegeG58cnpmAADABJUEkUI5vxFImSeVn3D44A1VoA5Ll4ea7AjhUHZRTOL7ZgzAYYPdcE9uDAAD7NVrtDpNU7nRZQlptTidL4-RFmChgOAeXC1BraAAqbUg6WI4BRsJWQxGJGwkhSm0kACVAuofB1ZgzjDAeaR5nyANJwKQbBBbApikCKZb9OlzCCLJBmalo7oKtYMiAAbVFUgAuotmZLJLj8WYsED9WLDTAKLapHz2aBcFz1Z1adrSHqDca+QKFobCVJGqT2OSxHAAGJgTWrem+iVTNkc93cpPYQNZ+YqmAdA0pqUyuXerPKgAUZgAdHXLDC0RQI1HKWBPQnFTqRWL5ugAJTIRZB2roHyBdL4czgaAwBwwFsQCmx+NE2XwMDzQmVyh4AjEUgmNAcbiOADynFalB4-fQUB3e8IDKPbGe58vuCPYD0jhjuFkmjXre96UBY1i2AgUCUMBO5gTYdhQXyqA3nesEgFY8GQS+vgBEERAeI4ABGcDsLIxCKLIdBAahlApGkGRttB6BAA

Things I don't know how I would handle is just A) ensuring that other locales have all the necessary params, B) the implications of having translation strings in TypeScript instead of say JSON files, and C) if handling stuff like plurals and stuff by just having more KV pairs would get annoying at all. But hey this would be pretty nice :)

**Comment by osdiab** [comment]
nm i see that this library already generates the proper params, nice :)

**Comment by Xiphe** [comment]
Wow! That also is a brilliant approach! Will have a deeper look later 👍

As for ensuring that all languages provide the same keys I think there are linters for that.

Here is the type declaration generator: https://github.com/Xiphe/typed-t

**Comment by Xiphe** [comment]
A) Ah, your question was about same params, not same keys (which should be solvable with your approach I guess) - in practice a language might not need all keys that another needs or maybe different ones, so maybe it would be better to have the params type require all parameters of all languages.


B) That was the main driver to my typed-t approach. I assume 99% of existing translations exist in the form of json and most saas translation management tools export to json.

C) Yeah, good question! I still want to explore providing typed-t for i18next but for my needs polyglot was just much more straightforward.

--

Wondering are you coming from a remix context? Asking because all of this really is not related to remix.

**Reply by osdiab** [reply]
Well if this issue https://github.com/microsoft/TypeScript/issues/32063 ever is resolved then it will be doable with JSON files :)

But yeah I am exploring now what's the best way to get good internationalization using Remix - we use NextJS at my company and there we use `next-translate` with some extra types to make it nice and typesafe, which works pretty nicely, but seems like nothing both stable and with great DX exists yet for Remix. Your approach looks the nicest but the reliance on internal Remix APIs is a little worrying lol. Hope the Remix maintainers put some love into this!

**Reply by Xiphe** [reply]
That's nice to hear! Thank you! I, too would love to get a more active discussion around this and hopefully contribute to a stable i18n solution with great performance and DX. Would appreciate if you (and maybe your coworkers) could chime in on https://github.com/remix-run/remix/discussions/3847 and https://github.com/remix-run/remix/discussions/3355 to give that a little more traction.

My next steps moving forward would be to try to bring some of the features of this tool to remix-i18next
